src/core/reference/include/openvino/reference/segment_max.hpp

```diff
@@ -10,47 +10,54 @@
 
 #include "openvino/core/shape.hpp"
 
-namespace ov {
-namespace reference {
+namespace ov::reference {
 
-template <typename T, typename T_idx>
+template <typename T,
+          typename T_idx,
+          typename std::enable_if<std::is_same<typename std::decay<T_idx>::type, int64_t>::value>::type* = nullptr>
 void segment_max(const T* data,
                  const Shape& data_shape,
                  const T_idx* segment_ids,
                  T* out,
                  const Shape& output_shape,
                  const T empty_segment_value) {
     const T_idx num_segments = output_shape[0];
-    const size_t inner_dim_size =
-        std::accumulate(data_shape.begin() + 1, data_shape.end(), 1, std::multiplies<size_t>());
-    std::vector<std::vector<T>> max_values(num_segments,
-                                           std::vector<T>(inner_dim_size, std::numeric_limits<T>::lowest()));
-    std::vector<bool> segment_has_values(num_segments, false);
+    const auto inner_dim_size = shape_size(data_shape.begin() + 1, data_shape.end());
+
+    // Initialize output with empty_segment_value
+    std::fill(out, out + num_segments * inner_dim_size, empty_segment_value);
 
     // Iterate over each element in the first dimension
-    T_idx base_idx;
     for (size_t i = 0; i < data_shape[0]; ++i) {
         const T_idx segment_id = segment_ids[i];
         if (segment_id >= num_segments) {
-            break;
+            continue;
         }
-        segment_has_values[segment_id] = true;
         // Iterate over each element in the inner dimensions
-        base_idx = i * inner_dim_size;
         for (size_t j = 0; j < inner_dim_size; ++j) {
+            const size_t index = i * inner_dim_size + j;
+            const size_t out_index = segment_id * inner_dim_size + j;
             // Update the maximum value for the current segment and inner dimension
-            max_values[segment_id][j] = std::max(max_values[segment_id][j], data[base_idx + j]);
+            out[out_index] = std::max(out[out_index], data[index]);
         }
     }
+}
 
-    // Populate the output array with the maximum values for each segment
-    for (T_idx segment_id = 0; segment_id < num_segments; ++segment_id) {
-        base_idx = segment_id * inner_dim_size;
-        for (size_t j = 0; j < inner_dim_size; ++j) {
-            out[base_idx + j] = segment_has_values[segment_id] ? max_values[segment_id][j] : empty_segment_value;
-        }
-    }
+template <typename T,
+          typename T_idx,
+          typename std::enable_if<!std::is_same<typename std::decay<T_idx>::type, int64_t>::value>::type* = nullptr>
```

Comment:
```suggestion
          std::enable_if_t<!std::is_same_v<std::decay_t<T_idx>, int64_t>>* = nullptr>
```
And same can be used for i64 version

Reply:
Applied.